polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.storage;
+
+import jakarta.validation.constraints.NotNull;
+import java.net.URI;
+import org.apache.polaris.core.storage.azure.AzureLocation;
+
+/** An abstraction over a storage location */
+public class StorageLocation {
+  private final String location;
+
+  public static StorageLocation of(String location) {
+    if (AzureLocation.isAzureLocation(location)) {
+      return new AzureLocation(location);
+    } else {
+      return new StorageLocation(location);
+    }
+  }
+
+  protected StorageLocation(@NotNull String location) {
+    if (location == null) {
+      this.location = null;
+    } else if (location.startsWith("file:/") && !location.startsWith("file:///")) {
+      this.location = URI.create(location.replaceFirst("file:/+", "file:///")).toString();
```

Comment:
This may not come up much in practice, but this isn't technically correct for `file://authority/path/to/file`.

The RFC for URIs is such that `file:/path/to/file` is equivalent to `file:///path/to/file`, just because the first case matches the grammar for unspecified authority, as does the last one as the canonical form.

Technically you're allowed to have file URIs like `file://remotehost:1234/path/to/file` which should *not* be turned into `file:///remotehost:1234/path/to/file`.

Reply:
Yeah, we should probably write an actual LocalStorageLocation to handle these, but I don't think this is an issue for now. Maybe, like you noted above, using Hadoop's `Path` can save us from this.